src/aiida_quantumespresso/workflows/xspectra/crystal.py

```diff
@@ -0,0 +1,676 @@
+
+        supercell = result.pop('supercell')
+        out_params = result.pop('output_parameters')
+        spacegroup_number = out_params['spacegroup_number']
+        if out_params.get_dict()['structure_is_standardized']:
+            standardized = result.pop('standardized_structure')
+            self.out('standardized_structure', standardized)
+        if spacegroup_number in range(1, 75): # trichoric system
+            self.ctx.eps_vectors = [[1., 0., 0.], [0., 1., 0.], [0., 0., 1.]]
+        if spacegroup_number in range(75, 195): # dichoric system
+            self.ctx.eps_vectors = [[1., 0., 0.], [0., 0., 1.]]
+        if spacegroup_number in range(195, 231): # isochoric system
+            self.ctx.eps_vectors = [[1., 0., 0.]]
+
+        structures_to_process = {f'{Key.split("_")[0]}_{Key.split("_")[1]}' : Value for Key, Value in result.items()}
+        self.ctx.structures_to_process = structures_to_process
+        self.ctx.equivalent_sites_data = out_params['equivalent_sites_data']
+
+        self.out('supercell_structure', supercell)
+        self.out('symmetry_analysis_data', out_params)
+
+    def should_run_upf2plotcore(self):
+        """If core wavefunction data files are specified, we skip the upf2plotcore step."""
+        return 'core_wfc_data' not in self.inputs
+
+    def run_upf2plotcore(self):
+        """Run the upf2plotcore.sh utility script for each element and return the core-wavefunction data."""
+
+        ShellJob = CalculationFactory('core.shell') # pylint: disable=invalid-name
+        elements_list = self.ctx.elements_list
+
+        shelljobs = {}
+        for element in elements_list:
+            upf = self.inputs.gipaw_pseudos[f'{element}']
+
+            shell_inputs = {}
+
+            shell_inputs['code'] = self.inputs.upf2plotcore_code
+            shell_inputs['nodes'] = {'upf': upf}
+            shell_inputs['arguments'] = ['upf']
+            shell_inputs['metadata'] = {'call_link_label': f'upf2plotcore_{element}'}
+
+            future_shelljob = self.submit(ShellJob, **shell_inputs)
+            self.report(f'Launching upf2plotcore.sh for {element}<{future_shelljob.pk}>')
+            shelljobs[f'upf2plotcore_{element}'] = future_shelljob
+
+        return ToContext(**shelljobs)
+
+
+    def inspect_upf2plotcore(self):
+        """Check that the outputs from the upf2plotcore step have yielded meaningful results.
+
+        This will simply check that the core wavefunction data returned contains at least
+        one core state and return an error if this is not the case.
+        """
+
+        labels = self.ctx.elements_list
+        for label in labels:
+            shelljob_node = self.ctx[f'upf2plotcore_{label}']
+            core_wfc_data = shelljob_node.outputs.stdout
+            header_line = core_wfc_data.get_content()[:40]
+            num_core_states = int(header_line.split(' ')[5])
+            if num_core_states == 0:
+                return self.exit_codes.ERROR_NO_GIPAW_INFO_FOUND
```

Comment:
The code here is repeated in the `XspectraCoreWorkChain`. Is it possible to avoid this?

Reply:
Bear in mind that this is used *instead* of the code used in the `CoreWorkChain`, thus only running upf2plotcore.sh once for each element before continuing to the `CoreWorkChain`s - thus saving a bit of computing effort.